File: 108-convert-sorted-array-to-binary-search-tree/108-convert-sorted-array-to-binary-search-tree.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

function sortedArrayToBST(nums: number[]): TreeNode | null {
    function getChild (nums, start, end){
        if(end < start) 
            return null;
        
        let mid = start + Math.floor((end - start) / 2);
        
        let node = new TreeNode(nums[mid]);
        node.left = getChild(nums, start, mid - 1);
        node.right = getChild(nums, mid+1, end);
        
        return node;
    }
    
    return getChild(nums, 0, nums.length - 1);
};

File: 108-convert-sorted-array-to-binary-search-tree/108-convert-sorted-array-to-binary-search-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def sortedArrayToBST(self, nums: List[int]) -> Optional[TreeNode]:
        def bst(num, start, end):
            if end < start:
                return None
            
            mid = start + (end- start)//2
            
            node = TreeNode(num[mid])
            node.left = bst(num, start, mid-1)
            node.right = bst(num, mid+1, end)
            return node
            
        return bst(nums, 0, len(nums)-1)
        
        